Kelompok Infolvators :
- M Rizqi Fadhilah
- M Arvin Fadriansyah
- Melliza Nastasia Izazi
- Thufael Bintang Alfattah
- Zulfikar fauzi
- Annisa Sulistyaningsih
- Niken Mustikaweni
- Galih refa

# **Youtube Views Prediction**
Salah satu implementasi regresi yang dapat dilakukan adalah dengan melakukan prediksi
views pada video youtube dengan menggunakan angka statistik atau atribut lain pada
videonya.

In [ ]:
# Melakukan Importing Library yang akan dibutuhkan nanti
import warnings
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error
from lazypredict.Supervised import LazyRegressor

Mengimpor berbagai library dan modul dari sklearn (scikit-learn) dan lazypredict yang akan digunakan untuk melakukan preprocessing data, pemodelan regresi, dan evaluasi model.

In [ ]:
# Mengabaikan semua peringatan
warnings.filterwarnings('ignore')

# Proses ekstraksi data
raw_data = pd.read_excel('youtube_statistics.xlsx')

# Tampilkan data
display(raw_data)

trending_date                                              title  \
0        2017-11-14  Sharry Mann: Cute Munda ( Song Teaser) | Parmi...   
1        2017-11-14  पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...   
2        2017-11-14  Stylish Star Allu Arjun @ ChaySam Wedding Rece...   
3        2017-11-14                     Eruma Saani | Tamil vs English   
4        2017-11-14  why Samantha became EMOTIONAL @ Samantha naga ...   
...             ...                                                ...   
36786    2018-06-14  फेकू आशिक़ - राजस्थान की सबसे शानदार कॉमेडी | ...   
36787    2018-06-14                         Seetha | Flowers | Ep# 364   
36788    2018-06-14  Bhramanam I Episode 87 - 12 June 2018 I Mazhav...   
36789    2018-06-14  Nua Bohu | Full Ep 285 | 13th June 2018 | Odia...   
36790    2018-06-14  Ee Nagaraniki Emaindi Trailer | Tharun Bhascke...   

            channel_title  category_id publish_time  \
0         Lokdhun Punjabi            1     12:20:39   
1                 HJ NEWS           25     05:43:56   
2                    TFPC           24     15:48:08   
3             Eruma Saani           23     07:08:48   
4              Filmylooks           24     01:14:16   
...                   ...          ...          ...   
36786      RDC Rajasthani           23     08:01:11   
36787          Flowers TV           24     11:30:04   
36788   Mazhavil Manorama           24     05:00:02   
36789           Tarang TV           24     15:07:49   
36790  Suresh Productions           24     04:29:54   

                                                    tags    views  likes  \
0      sharry mann|"sharry mann new song"|"sharry man...  1096327  33966   
1      पीरियड्स के समय|"पेट पर पति करता ऐसा"|"देखकर द...   590101    735   
2      Stylish Star Allu Arjun @ ChaySam Wedding Rece...   473988   2011   
3      Eruma Saani|"Tamil Comedy Videos"|"Films"|"Mov...  1242680  70353   
4      Filmylooks|"latest news"|"telugu movies"|"telu...   464015    492   
...                                                  ...      ...    ...   
36786  twinkle vaishnav comedy|"twinkle vaishnav"|"tw...   214378   3291   
36787  flowers serials|"actress"|"malayalam serials"|...   406828   1726   
36788  mazhavil manorama|"bhramanam full episode"|"gt...   386319   1216   
36789  tarang|"tarang tv"|"tarang tv online"|"tarang ...   130263    698   
36790  Ee Nagaraniki Emaindi|"Ee Nagaraniki Emaindi T...  1278249  22466   

       dislikes  comment_count  comments_disabled  ratings_disabled  \
0           798            882              False             False   
1           904              0               True             False   
2           243            149              False             False   
3          1624           2684              False             False   
4           293             66              False             False   
...         ...            ...                ...               ...   
36786       404            196              False             False   
36787       478           1428              False             False   
36788       453            697              False             False   
36789       115             65              False             False   
36790      1609           1205              False             False   

       video_error_or_removed  \
0                       False   
1                       False   
2                       False   
3                       False   
4                       False   
...                       ...   
36786                   False   
36787                   False   
36788                   False   
36789                   False   
36790                   False   

                                             description  No_tags  desc_len  \
0      Presenting Sharry Mann latest Punjabi Song  Cu...       15       920   
1      पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...       19      2232   
2      Watch Stylish Star Allu Arjun @ ChaySam Weddin...

In [ ]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36791 entries, 0 to 36790
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   trending_date           36791 non-null  object        
 1   title                   36791 non-null  object        
 2   channel_title           36791 non-null  object        
 3   category_id             36791 non-null  int64         
 4   publish_time            36791 non-null  object        
 5   tags                    36791 non-null  object        
 6   views                   36791 non-null  int64         
 7   likes                   36791 non-null  int64         
 8   dislikes                36791 non-null  int64         
 9   comment_count           36791 non-null  int64         
 10  comments_disabled       36791 non-null  bool          
 11  ratings_disabled        36791 non-null  bool          
 12  video_error_or_removed  36791 non-null  bool  

Menampilkan informasi tentang DataFrame raw_data.
DataFrame ini memiliki 36791 entri (baris) dan 18 kolom.
Setiap kolom memiliki tipe data yang berbeda-beda, seperti object, int64, bool, dan datetime64[ns].
Dan terdapat 45 missing value pada kolom deskcription.

# Transformasi Data

In [ ]:
col_dates = ['trending_date', 'publish_date']

for coldate in col_dates:
    raw_data[coldate] = pd.to_datetime(raw_data[coldate])

raw_data['publish_time'] = pd.to_timedelta(raw_data['publish_time'])

In [ ]:
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36791 entries, 0 to 36790
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype          
---  ------                  --------------  -----          
 0   trending_date           36791 non-null  datetime64[ns] 
 1   title                   36791 non-null  object         
 2   channel_title           36791 non-null  object         
 3   category_id             36791 non-null  int64          
 4   publish_time            36791 non-null  timedelta64[ns]
 5   tags                    36791 non-null  object         
 6   views                   36791 non-null  int64          
 7   likes                   36791 non-null  int64          
 8   dislikes                36791 non-null  int64          
 9   comment_count           36791 non-null  int64          
 10  comments_disabled       36791 non-null  bool           
 11  ratings_disabled        36791 non-null  bool           
 12  video_error_or_removed  36791 no

Mengubah tipe data kolom 'publish_time' menjadi tipe data **timedelta** sedangkan data kolom'trending_date' menjadi tipe data **datetime**
Tujuannya adalah untuk menyiapkan dataset dengan format tanggal dan waktu yang sesuai, agar dapat diproses dan dianalisis dengan lebih mudah.

# EDA

In [ ]:
# Periksa variabel kategorik
raw_data.select_dtypes(include = 'object')

title       channel_title  \
0      Sharry Mann: Cute Munda ( Song Teaser) | Parmi...     Lokdhun Punjabi   
1      पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...             HJ NEWS   
2      Stylish Star Allu Arjun @ ChaySam Wedding Rece...                TFPC   
3                         Eruma Saani | Tamil vs English         Eruma Saani   
4      why Samantha became EMOTIONAL @ Samantha naga ...          Filmylooks   
...                                                  ...                 ...   
36786  फेकू आशिक़ - राजस्थान की सबसे शानदार कॉमेडी | ...      RDC Rajasthani   
36787                         Seetha | Flowers | Ep# 364          Flowers TV   
36788  Bhramanam I Episode 87 - 12 June 2018 I Mazhav...   Mazhavil Manorama   
36789  Nua Bohu | Full Ep 285 | 13th June 2018 | Odia...           Tarang TV   
36790  Ee Nagaraniki Emaindi Trailer | Tharun Bhascke...  Suresh Productions   

                                                    tags  \
0      sharry mann|"sharry mann new song"|"sharry man...   
1      पीरियड्स के समय|"पेट पर पति करता ऐसा"|"देखकर द...   
2      Stylish Star Allu Arjun @ ChaySam Wedding Rece...   
3      Eruma Saani|"Tamil Comedy Videos"|"Films"|"Mov...   
4      Filmylooks|"latest news"|"telugu movies"|"telu...   
...                                                  ...   
36786  twinkle vaishnav comedy|"twinkle vaishnav"|"tw...   
36787  flowers serials|"actress"|"malayalam serials"|...   
36788  mazhavil manorama|"bhramanam full episode"|"gt...   
36789  tarang|"tarang tv"|"tarang tv online"|"tarang ...   
36790  Ee Nagaraniki Emaindi|"Ee Nagaraniki Emaindi T...   

                                             description  
0      Presenting Sharry Mann latest Punjabi Song  Cu...  
1      पीरियड्स के समय, पेट पर पति करता ऐसा, देखकर दं...  
2      Watch Stylish Star Allu Arjun @ ChaySam Weddin...  
3      This video showcases the difference between pe...  
4      why Samantha became EMOTIONAL @ Samantha naga ...  
...                                                  ...  
36786  PRG Music & RDC Rajasthani presents फेकू आशिक़...  
36787  Flowers - A R Rahman Show,Book your Tickets He...  
36788  Subscribe to Mazhavil Manorama now for your da...  
36789  Nuabohu : Story of a rustic village girl who w...  
36790  Check out Ee Nagaraniki Emaindi Trailer #EeNag...  

[36791 rows x 4 columns]

Menampilkan semua kolom dalam DataFrame raw_data yang memiliki tipe data 'object'. yang berguna untuk mengidentifikasi variabel-variabel kategorik dalam dataset

Karena sifatnya free text maka kita bisa abaikan fitur tersebut

In [ ]:
# Hapus kolom free text
raw_data = raw_data.drop(columns = ['title', 'channel_title', 'tags', 'description'])

In [ ]:
raw_data.head()

trending_date  category_id    publish_time    views  likes  dislikes  \
0    2017-11-14            1 0 days 12:20:39  1096327  33966       798   
1    2017-11-14           25 0 days 05:43:56   590101    735       904   
2    2017-11-14           24 0 days 15:48:08   473988   2011       243   
3    2017-11-14           23 0 days 07:08:48  1242680  70353      1624   
4    2017-11-14           24 0 days 01:14:16   464015    492       293   

   comment_count  comments_disabled  ratings_disabled  video_error_or_removed  \
0            882              False             False                   False   
1              0               True             False                   False   
2            149              False             False                   False   
3           2684              False             False                   False   
4             66              False             False                   False   

   No_tags  desc_len  len_title publish_date  
0       15       920         81   2017-11-12  
1       19      2232         58   2017-11-13  
2       14       482         58   2017-11-12  
3       20       263         30   2017-11-12  
4       11       753         88   2017-11-13

# Feature Engineering

In [ ]:
def categorize_time(td):
    hours = td.total_seconds() / 3600
    if (5 <= hours < 12):
        return 'morning'
    elif (12 <= hours < 17):
        return 'afternoon'
    elif (17 <= hours < 21):
        return 'evening'
    else:
        return 'night'



1.   hours = td.total_seconds() / 3600 menghitung jumlah jam dari parameter td (yang adalah tipe data timedelta).
2.  Fungsi ini menggunakan kondisi if-elif-else untuk mengkategorikan waktu berdasarkan jumlah jam:
* Jika jam antara 5 dan kurang dari 12, fungsi akan mengembalikan string 'morning'.
* Jika jam antara 12 dan kurang dari 17, fungsi akan mengembalikan string 'afternoon'.
* Jika jam antara 17 dan kurang dari 21, fungsi akan mengembalikan string 'evening'.
* Jika tidak masuk ke kategori di atas, fungsi akan mengembalikan string 'night'.

Fungsi ini berguna untuk mengkategorikan waktu publikasi suatu konten menjadi "pagi", "siang", "sore", atau "malam" berdasarkan jam publikasinya. Hal ini dapat membantu dalam analisis tren konten yang dipublikasikan pada waktu-waktu tertentu dalam sehari.



In [ ]:
raw_data['time_of_day'] = raw_data['publish_time'].apply(categorize_time)

raw_data.head()

trending_date  category_id    publish_time    views  likes  dislikes  \
0    2017-11-14            1 0 days 12:20:39  1096327  33966       798   
1    2017-11-14           25 0 days 05:43:56   590101    735       904   
2    2017-11-14           24 0 days 15:48:08   473988   2011       243   
3    2017-11-14           23 0 days 07:08:48  1242680  70353      1624   
4    2017-11-14           24 0 days 01:14:16   464015    492       293   

   comment_count  comments_disabled  ratings_disabled  video_error_or_removed  \
0            882              False             False                   False   
1              0               True             False                   False   
2            149              False             False                   False   
3           2684              False             False                   False   
4             66              False             False                   False   

   No_tags  desc_len  len_title publish_date time_of_day  
0       15       920         81   2017-11-12   afternoon  
1       19      2232         58   2017-11-13     morning  
2       14       482         58   2017-11-12   afternoon  
3       20       263         30   2017-11-12     morning  
4       11       753         88   2017-11-13       night

Menambahkan kolom baru 'time_of_day' ke dataset raw_data, yang mengkategorikan waktu publikasi video menjadi "morning", "afternoon, "evening", atau "night" berdasarkan jam publikasinya.

# Feature Selection
Mengidentifikasi fitur-fitur paling relevan yang dapat membantu dalam memahami karakteristik dan tren video pada dataset ini. Fitur-fitur ini dapat mewakili karakteristik video, seperti popularitas (views, likes, dislikes), interaksi pengguna (comment_count, comments_disabled, ratings_disabled), dan informasi teknis (video_error_or_removed, No_tags, desc_len, len_title).

In [ ]:
raw_data = raw_data.drop(columns = ['trending_date', 'category_id', 'publish_time', 'publish_date'])

raw_data.head()

views  likes  dislikes  comment_count  comments_disabled  \
0  1096327  33966       798            882              False   
1   590101    735       904              0               True   
2   473988   2011       243            149              False   
3  1242680  70353      1624           2684              False   
4   464015    492       293             66              False   

   ratings_disabled  video_error_or_removed  No_tags  desc_len  len_title  \
0             False                   False       15       920         81   
1             False                   False       19      2232         58   
2             False                   False       14       482         58   
3             False                   False       20       263         30   
4             False                   False       11       753         88   

  time_of_day  
0   afternoon  
1     morning  
2   afternoon  
3     morning  
4       night

Menyederhanakan dataset dengan hanya mempertahankan kolom-kolom yang relevan untuk analisis lebih lanjut, seperti metrik terkait video **(views, likes, dislikes, dll.)**, informasi tentang video **(No_tags, desc_len, len_title)**, dan kategori waktu publikasi** (time_of_day)**.

#Encoding

In [ ]:
col_encode = ['comments_disabled', 'ratings_disabled', 'video_error_or_removed']

map_boolean = {
    False : 0,
    True : 1
}

for col_enc in col_encode:
    raw_data[col_enc] = raw_data[col_enc].map(map_boolean)

1. **col_encode = ['comments_disabled', 'ratings_disabled', 'video_error_or_removed']:**

    * menentukan kolom-kolom yang akan di-encode menggunakan One-Hot Encoding.
2. **map_boolean = { 'False': 0, 'True': 1 }**:
    * mendefinisikan sebuah pemetaan (dictionary) untuk mengubah nilai boolean 'False' menjadi 0 dan 'True' menjadi 1.
3. **for col_enc in col_encode:**
    * merupakan loop yang akan berjalan untuk setiap kolom yang terdapat dalam col_encode.
4. raw_data[col_enc] = raw_data[col_enc].map(map_boolean):
    * pemetakan menggunakan map_boolean dictionary pada setiap kolom yang terdapat dalam col_encode.
yang akan mengubah nilai boolean pada kolom-kolom tersebut menjadi nilai numerik 0 dan 1.

Secara singkat, kode ini melakukan One-Hot Encoding pada kolom-kolom tertentu untuk mengubah data kategorikal menjadi format numerikal.

In [ ]:
from sklearn.preprocessing import LabelEncoder

# Inisialisasi LabelEncoder
le = LabelEncoder()

# Fit dan transform kolom kategorik
raw_data['time_of_day'] = le.fit_transform(raw_data['time_of_day'])

# Tampilkan
raw_data.head()

views  likes  dislikes  comment_count  comments_disabled  \
0  1096327  33966       798            882                  0   
1   590101    735       904              0                  1   
2   473988   2011       243            149                  0   
3  1242680  70353      1624           2684                  0   
4   464015    492       293             66                  0   

   ratings_disabled  video_error_or_removed  No_tags  desc_len  len_title  \
0                 0                       0       15       920         81   
1                 0                       0       19      2232         58   
2                 0                       0       14       482         58   
3                 0                       0       20       263         30   
4                 0                       0       11       753         88   

   time_of_day  
0            0  
1            2  
2            0  
3            2  
4            3

1. from sklearn.preprocessing import **LabelEncoder**:
    * mengimport kelas **LabelEncoder** dari modul **sklearn.preprocessing**.
LabelEncoder digunakan untuk mengkodekan fitur kategorikal menjadi nilai numerik.
2. Inisialisasi LabelEncoder:
    * menginisialisasi objek LabelEncoder dengan perintah **le = LabelEncoder()**.
3. Fit dan transform kolom kategorik:
    * Baris **raw_data['time_of_day'] = le.fit_transform(raw_data['time_of_day'])** melakukan transformasi pada kolom 'time_of_day' menggunakan LabelEncoder.
Transformasi ini akan mengubah nilai kategorikal pada kolom tersebut menjadi nilai numerik.
4. Tampilkan:
Perintah **raw_data.head()** akan menampilkan beberapa baris awal dari dataset raw_data.

Secara singkat, kode ini melakukan Label Encoding pada kolom 'time_of_day' dalam dataset raw_data. Proses ini bertujuan untuk mengubah nilai kategorikal menjadi nilai numerik agar dapat diproses oleh model pembelajaran mesin.

# Splitting Data

In [ ]:
X = raw_data.drop(columns = ['views'])
y = raw_data[['views']]

`X`: DataFrame yang berisi fitur-fitur **(variabel independen)** yang akan digunakan sebagai input ke dalam model. **Kolom 'views' telah dihapus dari X**.

`y`: DataFrame yang hanya berisi **kolom target 'views'**, yang merupakan nilai yang ingin diprediksi oleh model.

In [ ]:
X

likes  dislikes  comment_count  comments_disabled  ratings_disabled  \
0      33966       798            882                  0                 0   
1        735       904              0                  1                 0   
2       2011       243            149                  0                 0   
3      70353      1624           2684                  0                 0   
4        492       293             66                  0                 0   
...      ...       ...            ...                ...               ...   
36786   3291       404            196                  0                 0   
36787   1726       478           1428                  0                 0   
36788   1216       453            697                  0                 0   
36789    698       115             65                  0                 0   
36790  22466      1609           1205                  0                 0   

       video_error_or_removed  No_tags  desc_len  len_title  time_of_day  
0                           0       15       920         81            0  
1                           0       19      2232         58            2  
2                           0       14       482         58            0  
3                           0       20       263         30            2  
4                           0       11       753         88            3  
...                       ...      ...       ...        ...          ...  
36786                       0       19       818        100            2  
36787                       0        8       468         26            2  
36788                       0        8      2187         57            2  
36789                       0       15       432         64            0  
36790                       0       14      1030         76            3  

[36791 rows x 10 columns]

Untuk menampilkan Nilai X yang tidak memiliki kolom views

Setiap kolom dalam DataFrame X mewakili fitur yang digunakan untuk memprediksi views:

1. likes: Jumlah likes yang diterima oleh video.
2. dislikes: Jumlah dislikes yang diterima oleh video.
3. comment_count: Jumlah komentar yang diterima oleh video.
4. comments_disabled: Status apakah komentar pada video dinonaktifkan (0 berarti tidak dinonaktifkan, 1 berarti dinonaktifkan).
5. ratings_disabled: Status apakah rating pada video dinonaktifkan (0 berarti tidak dinonaktifkan, 1 berarti dinonaktifkan).
6. video_error_or_removed: Status apakah video memiliki error atau dihapus (0 berarti tidak, 1 berarti ya).
7. No_tags: Jumlah tag yang digunakan dalam video.
8. desc_len: Panjang deskripsi video dalam jumlah karakter.
9. len_title: Panjang judul video dalam jumlah karakter.
10. time_of_day: Waktu dalam hari saat video diunggah (dinyatakan dalam bilangan bulat yang mewakili jam).

In [ ]:
y

views
0      1096327
1       590101
2       473988
3      1242680
4       464015
...        ...
36786   214378
36787   406828
36788   386319
36789   130263
36790  1278249

[36791 rows x 1 columns]

Menampilkan kolom y yang berisi kolom views yang sudah dipisah sebelumnya.

views: Kolom ini mencatat jumlah penayangan untuk setiap video. Nilai ini akan digunakan sebagai target dalam model prediksi. Model akan dilatih untuk memprediksi nilai views berdasarkan fitur-fitur yang ada di variabel X. DataFrame y hanya memiliki satu kolom yaitu views, yang berisi jumlah penayangan untuk setiap video. Ada total 36791 baris, yang berarti ada 36791 video yang diamati dalam dataset ini.

In [ ]:
from sklearn.model_selection import train_test_split

# Bagi data menjadi training dan testing set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

Melakukan pembagian Data Training dan Testing set yaitu :      
1. Test_size = 0.3 yang berarti dari data tersebut akan diambil sekitar 30% dari total data pada `youtube_statistic`
2. Training = Maka sisa atau 70% dari data tersebut akan menjadi data training.

In [ ]:
print(f'Total data awal {raw_data.shape}')
print(f'Jumlah data latih : {X_train.shape} dan jumlah data test : {X_test.shape}')

Total data awal (36791, 11)
Jumlah data latih : (25753, 10) dan jumlah data test : (11038, 10)


Dari data yang didapatkan Total data awal dari Dataset `youtube_statistic` terdapat **36791 data dan 11 kolom**, yang kita bagi menjadi **menjadi 7 banding 3**

dimana 7 merupakan data training : **Jumlah data latih : (25753, 10)** , dan 3 pada data testing : **jumlah data test : (11038, 10)**

# Scaling

In [ ]:
from sklearn.preprocessing import MinMaxScaler

# Rentang default adalah [0, 1]
scaler = MinMaxScaler()

# Fit dan transform data
scaled_train = scaler.fit_transform(X_train)

# Buat DataFrame dari data yang telah diskalakan
X_train_scaled = pd.DataFrame(scaled_train, columns=X_train.columns)

X_train_scaled

likes  dislikes  comment_count  comments_disabled  ratings_disabled  \
0       0.00      0.00           0.00               1.00              0.00   
1       0.00      0.00           0.00               0.00              0.00   
2       0.01      0.00           0.00               0.00              0.00   
3       0.01      0.00           0.01               0.00              0.00   
4       0.00      0.00           0.00               0.00              0.00   
...      ...       ...            ...                ...               ...   
25748   0.00      0.00           0.00               0.00              0.00   
25749   0.00      0.00           0.00               0.00              0.00   
25750   0.00      0.00           0.00               0.00              0.00   
25751   0.00      0.00           0.00               0.00              0.00   
25752   0.00      0.00           0.00               0.00              1.00   

       video_error_or_removed  No_tags  desc_len  len_title  time_of_day  
0                        0.00     0.00      0.02       0.49         1.00  
1                        0.00     0.31      0.07       0.87         0.67  
2                        0.00     0.24      0.16       0.75         0.33  
3                        0.00     0.42      0.06       0.48         0.33  
4                        0.00     0.17      0.11       0.77         0.67  
...                       ...      ...       ...        ...          ...  
25748                    0.00     0.31      0.06       0.25         0.67  
25749                    0.00     0.13      0.06       0.31         0.33  
25750                    0.00     0.24      0.19       0.43         0.00  
25751                    0.00     0.28      0.22       0.98         0.00  
25752                    0.00     0.08      0.10       0.85         0.67  

[25753 rows x 10 columns]

Kode di atas menunjukkan proses normalisasi data menggunakan MinMaxScaler pada set data pelatihan (X_train). Data dalam X_train_scaled menunjukkan bahwa semua fitur telah dinormalisasi ke dalam rentang [0, 1]. Contoh data yang ditampilkan menunjukkan nilai-nilai yang telah dinormalisasi untuk beberapa fitur seperti likes, dislikes, comment_count, dan lainnya.

Normalisasi ini penting untuk memastikan bahwa semua fitur berkontribusi secara proporsional dalam model pembelajaran mesin dan tidak ada fitur yang mendominasi karena rentang nilainya yang lebih besar.

In [ ]:
# Fit dan transform data
scaled_y_train = scaler.fit_transform(y_train)

# Buat DataFrame dari data yang telah diskalakan
y_train_scaled = pd.DataFrame(scaled_y_train, columns=y_train.columns)

y_train_scaled

views
0       0.00
1       0.00
2       0.03
3       0.04
4       0.00
...      ...
25748   0.00
25749   0.00
25750   0.00
25751   0.00
25752   0.03

[25753 rows x 1 columns]

**1.Fit dan Transform Data:**


scaled_y_train = scaler.fit_transform(y_train):
Melakukan fitting scaler pada data target (y_train) dan kemudian mentransformasi data tersebut. Proses fitting menghitung nilai minimum dan maksimum dari y_train, dan transformasi mengubah nilai-nilai tersebut ke dalam rentang [0, 1].

**2.Membuat DataFrame Baru:**


y_train_scaled = pd.DataFrame(scaled_y_train, columns=y_train.columns):
Membuat DataFrame y_train_scaled dari data yang telah diskalakan (scaled_y_train), dengan mempertahankan nama kolom yang sama seperti pada y_train.

**3.Hasil Akhir:**


Data dalam y_train_scaled menunjukkan bahwa semua nilai target (views) telah dinormalisasi ke dalam rentang [0, 1]. Ini berarti nilai terkecil diubah menjadi 0 dan nilai terbesar diubah menjadi 1, dengan nilai lainnya disesuaikan secara proporsional di antara 0 dan 1.

In [ ]:
from sklearn.preprocessing import MinMaxScaler

# Rentang default adalah [0, 1]
scaler = MinMaxScaler()

# Fit dan transform data
scaled_test = scaler.fit_transform(X_test)
scaled_y_test = scaler.fit_transform(y_test)

# Buat DataFrame dari data yang telah diskalakan
X_test_scaled = pd.DataFrame(scaled_test, columns=X_test.columns)
y_test_scaled = pd.DataFrame(scaled_y_test, columns=y_test.columns)

X_test_scaled

likes  dislikes  comment_count  comments_disabled  ratings_disabled  \
0       0.00      0.00           0.00               0.00              0.00   
1       0.00      0.00           0.00               0.00              0.00   
2       0.00      0.00           0.00               0.00              0.00   
3       0.00      0.00           0.00               0.00              0.00   
4       0.00      0.00           0.00               0.00              0.00   
...      ...       ...            ...                ...               ...   
11033   0.00      0.00           0.00               0.00              0.00   
11034   0.01      0.00           0.00               0.00              0.00   
11035   0.00      0.00           0.00               0.00              0.00   
11036   0.00      0.00           0.00               0.00              0.00   
11037   0.00      0.00           0.00               1.00              0.00   

       video_error_or_removed  No_tags  desc_len  len_title  time_of_day  
0                        0.00     0.30      0.20       0.94         0.67  
1                        0.00     0.10      0.02       0.83         0.33  
2                        0.00     0.08      0.06       0.67         0.00  
3                        0.00     0.25      0.33       0.94         1.00  
4                        0.00     0.27      0.21       1.00         0.67  
...                       ...      ...       ...        ...          ...  
11033                    0.00     0.24      0.31       0.53         0.00  
11034                    0.00     0.27      0.59       0.12         0.33  
11035                    0.00     0.30      0.49       0.98         1.00  
11036                    0.00     0.15      0.07       0.27         0.33  
11037                    0.00     0.13      0.09       0.69         0.33  

[11038 rows x 10 columns]

Penjelasan:

- from sklearn.preprocessing import MinMaxScaler untuk mengubah skala data numerik ke rentang yang ditentukan

- scaler = MinMaxScaler()
 MinMaxScaler dibuat dengan rentang default [0, 1].

- scaled_test = scaler.fit_transform(X_test)
scaled_y_test = scaler.fit_transform(y_test)
digunakan untuk menyesuaikan skala dengan data dan kemudian mentransformasi data tersebut. Dalam hal ini, X_test dan y_test adalah dataset yang akan diskalakan.


- X_test_scaled = pd.DataFrame(scaled_test, columns=X_test.columns)
y_test_scaled = pd.DataFrame(scaled_y_test, columns=y_test.columns)
Setelah data diskalakan, hasilnya dikonversi kembali menjadi DataFrame Pandas dengan kolom yang sama seperti dataset asli.

- X_test_scaled
Menampilkan DataFrame X_test_scaled yang berisi data X_test yang sudah diskalakan.

In [ ]:
# Ubah y value menjadi 1 dimensi
y_train_scaled = y_train_scaled.to_numpy().ravel()
y_test_scaled = y_test_scaled.to_numpy().ravel()

Mengubah nilai y yang tadinya 2 dimensi menjadi 1 dimensi.

In [ ]:
print(X_train_scaled.shape)
print(X_test_scaled.shape)
print(y_train_scaled.shape)
print(y_test_scaled.shape)


(25753, 10)
(11038, 10)
(25753,)
(11038,)


# Modeling

In [ ]:
# Cari Algoritma regresi terbaik
reg = LazyRegressor(verbose=0, custom_metric=None, predictions=True)
models, predictions = reg.fit(X_train_scaled, X_test_scaled, y_train_scaled, y_test_scaled)
models

100%|██████████| 42/42 [08:34<00:00, 12.25s/it]


Adjusted R-Squared      R-Squared    RMSE  \
Model                                                                      
ExtraTreesRegressor                          0.97           0.97    0.00   
RandomForestRegressor                        0.95           0.95    0.01   
XGBRegressor                                 0.95           0.95    0.01   
BaggingRegressor                             0.94           0.94    0.01   
DecisionTreeRegressor                        0.90           0.91    0.01   
GradientBoostingRegressor                    0.89           0.89    0.01   
KNeighborsRegressor                          0.89           0.89    0.01   
NuSVR                                        0.89           0.89    0.01   
LGBMRegressor                                0.88           0.88    0.01   
ExtraTreeRegressor                           0.87           0.87    0.01   
HistGradientBoostingRegressor                0.86           0.86    0.01   
ElasticNetCV                                 0.76           0.76    0.01   
LassoCV                                      0.76           0.76    0.01   
RidgeCV                                      0.76           0.76    0.01   
BayesianRidge                                0.76           0.76    0.01   
Ridge                                        0.76           0.76    0.01   
LassoLarsCV                                  0.76           0.76    0.01   
LassoLarsIC                                  0.76           0.76    0.01   
LarsCV                                       0.76           0.76    0.01   
Lars                                         0.76           0.76    0.01   
TransformedTargetRegressor                   0.76           0.76    0.01   
LinearRegression                             0.76           0.76    0.01   
OrthogonalMatchingPursuitCV                  0.75           0.76    0.01   
AdaBoostRegressor                            0.75           0.75    0.01   
OrthogonalMatchingPursuit                    0.71           0.71    0.01   
TweedieRegressor                             0.64           0.64    0.02   
HuberRegressor                               0.62           0.62    0.02   
MLPRegressor                                 0.18           0.18    0.02   
PoissonRegressor                             0.07           0.07    0.03   
LinearSVR                                    0.06           0.06    0.03   
DummyRegressor                              -0.00          -0.00    0.03   
LassoLars                                   -0.00          -0.00    0.03   
ElasticNet                                  -0.00          -0.00    0.03   
Lasso                                       -0.00          -0.00    0.03   
RANSACRegressor                             -0.63          -0.63    0.03   
PassiveAggressiveRegressor                  -4.41          -4.40    0.06   
SVR                                         -5.92          -5.91    0.07   
SGDRegressor                       -6144578114.51 -6139010860.62 2032.34   

                               Time Taken  
Model                                      
ExtraTreesRegressor                 10.81  
RandomForestRegressor               21.94  
XGBRegressor                         1.14  
BaggingRegressor                     2.62  
DecisionTreeRegressor                0.39  
GradientBoostingRegressor            5.70  
KNeighborsRegressor                  2.32  
NuSVR                              276.28  
LGBMRegressor                        0.20  
ExtraTreeRegressor                   0.18  
HistGradientBoostingRegressor        1.28  
ElasticNetCV                         0.51  
LassoCV                              0.45  
RidgeCV                              0.05  
BayesianRidge                        0.07  
Ridge                                0.03  
LassoLarsCV                          0.17  
LassoLarsIC                          0.12  
LarsCV                               0.18  
Lars                                 0.35  
TransformedTargetRegre

**Terbaik Secara Keseluruhan**: ExtraTreesRegressor – Memiliki Adjusted R-Squared dan R-Squared tertinggi, serta RMSE terendah.

**Peringkat Kedua**: RandomForestRegressor – Performa mendekati ExtraTreesRegressor tetapi memerlukan waktu lebih lama untuk dijalankan.

**Performa Cepat yang Baik**: XGBRegressor – Metrik yang sebanding tetapi waktu eksekusi secara signifikan lebih cepat.

ExtraTreesRegressor umumnya dianggap sebagai model terbaik berdasarkan hasil ini, menyeimbangkan performa dan waktu komputasi.

In [ ]:
from sklearn.ensemble import ExtraTreesRegressor
from sklearn.model_selection import GridSearchCV, KFold

extraTreeRegressor = ExtraTreesRegressor()
cv = KFold(random_state=123, shuffle=True)
param_grid = {'criterion': ['squared_error', 'friedman_mse'],
              'min_samples_split': [2, 5, 10, 15, 20, 25],
              'max_depth' : [5, 10, 15, None]}
clf2 = GridSearchCV(estimator=extraTreeRegressor, param_grid=param_grid, cv=cv)

clf2.fit(X_train_scaled, y_train_scaled)
clf2.best_params_

{'criterion': 'squared_error', 'max_depth': None, 'min_samples_split': 2}

**ExtraTreesRegressor**: Sebuah model machine learning yang membangun beberapa decision tree dan merata-rata hasil prediksinya untuk meningkatkan akurasi dan mengurangi overfitting.
**GridSearchCV**: Ini adalah alat yang melakukan pencarian menyeluruh (grid search) untuk kombinasi hyperparameter terbaik berdasarkan kriteria evaluasi tertentu (misalnya, akurasi, MSE). GridSearchCV mencoba semua kombinasi yang mungkin dari hyperparameter yang diberikan.
**KFold**: Ini adalah metode cross-validation yang membagi data menjadi beberapa subset (folds). Model dilatih pada beberapa subset ini dan divalidasi pada subset lain, yang memastikan bahwa model diuji pada berbagai bagian data yang berbeda.

**param_grid**: Ini adalah dictionary yang menentukan hyperparameter yang akan diuji. Setiap hyperparameter memiliki beberapa opsi yang akan dicoba oleh GridSearchCV.
criterion: Menentukan fungsi yang digunakan untuk mengukur kualitas split di setiap node pohon.
**squared_error**: Mean Squared Error standar, yang menghitung kuadrat dari selisih antara nilai yang diprediksi dan nilai sebenarnya.
**friedman_mse**: Variasi dari MSE yang dioptimalkan untuk pohon regresi, dapat memberikan hasil yang lebih baik dalam situasi tertentu.
**min_samples_split**: Menentukan jumlah minimum sampel yang diperlukan untuk membagi node internal. Nilai yang lebih besar membuat pohon lebih konservatif, menghindari overfitting.
Contoh: min_samples_split=2 berarti node akan dibagi jika ada setidaknya 2 sampel.
**max_depth**: Menentukan kedalaman maksimum pohon. Batasan kedalaman pohon membantu mencegah pohon menjadi terlalu kompleks, yang bisa mengakibatkan overfitting.
Contoh: max_depth=10 berarti pohon akan dibatasi hingga 10 tingkat.

**clf2.best_params_**: Setelah model dilatih dengan berbagai kombinasi hyperparameter, kode ini akan mengembalikan kombinasi hyperparameter yang memberikan performa terbaik berdasarkan evaluasi cross-validation.

# Prediksi

In [ ]:
etr = ExtraTreesRegressor(**clf2.best_params_)
etr.fit(X_train_scaled, y_train_scaled)
y_pred = etr.predict(X_test_scaled)
y_pred

array([0.00261027, 0.00606368, 0.00253015, ..., 0.0045541 , 0.004891  ,
       0.00465161])

Kode dilatih dengan menggunakan hyperparameter clf2.best_param
kemudian model diuji ke x_test_scaled untuk mengetahui apakah model sudah akurat atau belum yang akan diuji di kose selanjutnya

# Evaluasi

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

MAE = mean_absolute_error(y_test_scaled, y_pred)
MSE = mean_squared_error(y_test_scaled, y_pred)
RMSE = np.sqrt(mean_squared_error(y_test_scaled, y_pred))
R_Squared = r2_score(y_test_scaled, y_pred)

print(f'Error MAE = {MAE}')
print(f'Error MSE = {MSE}')
print(f'RMSE = {RMSE}')
print(f'R-Squared = {R_Squared}')

Error MAE = 0.0018160574039167861
Error MSE = 2.091150892761421e-05
RMSE = 0.004572910334525947
R-Squared = 0.9689192049202318


**MAE**: Rata-rata dari perbedaan absolut antara nilai aktual (y_test_scaled) dan nilai prediksi (y_pred). MAE memberikan gambaran seberapa jauh prediksi berada dari nilai sebenarnya, tanpa mempertimbangkan arah kesalahan (apakah terlalu tinggi atau terlalu rendah).

**MSE**: Rata-rata dari kuadrat perbedaan antara nilai aktual dan nilai prediksi. MSE menekankan kesalahan yang lebih besar karena mengkuadratkan selisih, sehingga model yang menghasilkan prediksi dengan kesalahan besar akan memiliki nilai MSE yang lebih tinggi.

**RMSE**: Akar kuadrat dari MSE. RMSE berada dalam satuan yang sama dengan variabel target (dalam kasus ini, data yang telah di-scaled), sehingga lebih mudah diinterpretasikan daripada MSE. RMSE memberikan bobot lebih pada kesalahan besar, mirip dengan MSE, namun lebih mudah diinterpretasikan karena berada dalam skala yang sama dengan data.

**R-Squared**: Mengukur proporsi variasi dalam variabel target yang dapat dijelaskan oleh model. Nilai R² berkisar antara 0 dan 1, dengan 1 menunjukkan bahwa model sempurna dalam menjelaskan semua variasi dalam data target, dan 0 menunjukkan bahwa model tidak lebih baik daripada sekadar mengambil rata-rata dari target.